src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1490 @@
+
+    /** Maps symbols of all methods to their corresponding declarations.
+     */
+    private final Map<Symbol, MethodInfo> methodMap = new LinkedHashMap<>();
+
+    /** Contains symbols of fields and constructors that have warnings suppressed.
+     */
+    private final Set<Symbol> suppressed = new HashSet<>();
+
+    /** The declaring class of the constructor we're currently analyzing.
+     *  This is the 'this' type we're trying to detect leaks of.
+     */
+    private JCClassDecl targetClass;
+
+    /** Snapshots of {@link #callStack} where possible 'this' escapes occur.
+     */
+    private final ArrayList<DiagnosticPosition[]> warningList = new ArrayList<>();
+
+// These fields are scoped to the constructor being analyzed
+
+    /** The declaring class of the "invoked" method we're currently analyzing.
+     *  This is either the analyzed constructor or some method it invokes.
+     */
+    private JCClassDecl methodClass;
+
+    /** The current "call stack" during our analysis. The first entry is some method
+     *  invoked from the target constructor; if empty, we're still in the constructor.
+     */
+    private final ArrayDeque<DiagnosticPosition> callStack = new ArrayDeque<>();
+
+    /** Used to terminate recursion in {@link #invokeInvokable invokeInvokable()}.
+     */
+    private final Set<Pair<JCTree, RefSet<Ref>>> invocations = new HashSet<>();
+
+    /** Snapshot of {@link #callStack} where a possible 'this' escape occurs.
+     *  If non-null, a 'this' escape warning has been found in the current
+     *  constructor statement, initialization block statement, or field initializer.
+     */
+    private DiagnosticPosition[] pendingWarning;
+
+// These fields are scoped to the constructor or invoked method being analyzed
+
+    /** Current lexical scope depth in the constructor or method we're currently analyzing.
+     *  Depth zero is the outermost scope. Depth -1 means we're not analyzing.
+     */
+    private int depth = -1;
+
+    /** Possible 'this' references in the constructor or method we're currently analyzing.
+     *  Null value means we're not analyzing.
+     */
+    private RefSet<Ref> refs;
+
+// Constructor
+
+    ThisEscapeAnalyzer(Names names, Symtab syms, Types types, Log log, Lint lint) {
+        this.names = names;
+        this.syms = syms;
+        this.types = types;
+        this.log = log;
+        this.lint = lint;
+    }
+
+//
+// Main method
+//
+
+    public void analyzeTree(Env<AttrContext> env) {
+
+        // Sanity check
+        Assert.check(checkInvariants(false, false));
+
+        // Short circuit if warnings are totally disabled
+        if (!lint.isEnabled(Lint.LintCategory.THIS_ESCAPE))
+            return;
+
+        // Build a mapping from symbols of methods to their declarations.
+        // Classify all ctors and methods as analyzable and/or invokable.
+        // Track which constructors and fields have warnings suppressed.
+        new TreeScanner() {
```

Comment:
I wonder if it would be better to have this visitor be defined externally, the benefit of this would be that it can be created only once instead of every time we enter this method, also I think it would improve the readability of this method. But I let this to your consideration

Reply:
This method is only supposed to ever be entered once...  but you made me realize that there should probably be an assertion check for that assumption.